Neos.Neos/Classes/Fusion/Cache/GraphProjectorCatchUpHookForCacheFlushing.php

```diff
@@ -173,66 +222,104 @@ public function onAfterEvent(EventInterface $eventInstance, EventEnvelope $event
-     */
-    private array $cacheFlushesOnAfterCatchUp = [];
-
-    protected function scheduleCacheFlushJobForNodeAggregate(
+    private function scheduleCacheFlushJobForNodeAggregate(
         ContentRepository $contentRepository,
         WorkspaceName $workspaceName,
-        NodeAggregateId $nodeAggregateId
+        NodeAggregate $nodeAggregate
     ): void {
         // we store this in an associative array deduplicate.
-        $this->cacheFlushesOnBeforeBatchCompleted[$workspaceName->value . '__' . $nodeAggregateId->value] = [
-            'cr' => $contentRepository,
-            'wsn' => $workspaceName,
-            'nai' => $nodeAggregateId
-        ];
+        $this->flushNodeAggregateRequestsOnBeforeBatchCompleted[$workspaceName->value . '__' . $nodeAggregate->nodeAggregateId->value] = FlushNodeAggregateRequest::create(
+            $contentRepository->id,
+            $workspaceName,
+            $nodeAggregate->nodeAggregateId,
+            $nodeAggregate->nodeTypeName,
+            $this->determineParentNodeAggregateIds($workspaceName, $nodeAggregate->nodeAggregateId, NodeAggregateIds::createEmpty())
+        );
     }
 
-    public function onBeforeBatchCompleted(): void
+    private function scheduleCacheFlushJobForWorkspaceName(
+        ContentRepository $contentRepository,
+        WorkspaceName $workspaceName
+    ): void {
+        // we store this in an associative array deduplicate.
+        $this->flushWorkspaceRequestsOnBeforeBatchCompleted[$workspaceName->value] = FlushWorkspaceRequest::create(
+            $contentRepository->id,
+            $workspaceName,
+        );
+    }
+
+    private function determineParentNodeAggregateIds(WorkspaceName $workspaceName, NodeAggregateId $childNodeAggregateId, NodeAggregateIds $collectedParentNodeAggregateIds): NodeAggregateIds
     {
-        foreach ($this->cacheFlushesOnBeforeBatchCompleted as $k => $entry) {
-            $this->contentCacheFlusher->flushNodeAggregate($entry['cr'], $entry['wsn'], $entry['nai']);
-            $this->cacheFlushesOnAfterCatchUp[$k] = $entry;
+        $parentNodeAggregates = $this->contentRepository->getContentGraph($workspaceName)->findParentNodeAggregates($childNodeAggregateId);
+        $parentNodeAggregateIds = NodeAggregateIds::fromArray(
+            array_map(static fn (NodeAggregate $parentNodeAggregate) => $parentNodeAggregate->nodeAggregateId, iterator_to_array($parentNodeAggregates))
+        );
+
+        foreach ($parentNodeAggregateIds as $parentNodeAggregateId) {
+            // Prevent infinite loops
+            if (!$collectedParentNodeAggregateIds->contain($parentNodeAggregateId)) {
+                $collectedParentNodeAggregateIds = $collectedParentNodeAggregateIds->merge(NodeAggregateIds::create($parentNodeAggregateId));
+                $collectedParentNodeAggregateIds = $this->determineParentNodeAggregateIds($workspaceName, $parentNodeAggregateId, $collectedParentNodeAggregateIds);
```

Comment:
We need to prevent infinite loops here, but just for a theroretical case. AFAIR we do not allow circular dependencies, BUT we have a test case which creates that case to test, that it doesn't happen. 🤷‍♂️